File: csharp/sieve/Sieve.cs
using System;
using System.Collections;
using System.Collections.Generic;

public static class Sieve
{
    public static IEnumerable<int> Primes(int max)
    {
        if (max < 0)
            throw new ArgumentOutOfRangeException(nameof(max));

        var primes = new BitArray(max + 1, true);

        for (var i = 2; i <= max; i++)
        {
            if (!primes[i])
                continue;

            for (var j = i * 2; j <= max; j += i)
                primes[j] = false;

            yield return i;
        }
    }
}

File: ruby/sieve/sieve.rb
class Sieve
  def initialize(max) = @max = max

  def primes
    sieve = Array.new(@max + 1, true)

    (2..@max).each do |number|
      next unless sieve[number]

      (number * 2..@max).step(number).each { |multiple| sieve[multiple] = false }
    end

    sieve.each_with_index.filter_map { |is_prime, i| i if is_prime && i > 1 }
  end
end